File: src/main/java/dev/klomptech/jbark/parser/collection/TakeParser.java
package dev.klomptech.jbark.parser.collection;

import dev.klomptech.jbark.config.ConfigLoader;
import dev.klomptech.jbark.errors.BarkError;
import dev.klomptech.jbark.lexer.Token;
import dev.klomptech.jbark.lexer.TokenType;
import dev.klomptech.jbark.parser.AstNode;
import dev.klomptech.jbark.parser.Keywords;
import dev.klomptech.jbark.parser.ParseExpression;
import dev.klomptech.jbark.parser.Parser;
import dev.klomptech.jbark.print.PrintStyle;
import java.util.ArrayList;
import java.util.List;

public class TakeParser {

  // Which stash and which item word (first, second, last, etc.)
  private record StashTake(String stash, String which) {}

  private final Parser parser;

  public TakeParser(final Parser parser) {
    this.parser = parser;
  }

  // Dog takes one item from a stash (example: she wants the first biscuit from her cookie jar)
  public AstNode parseTakeLine() throws BarkError {
    // Look for the stash and the item word
    StashTake take = findStashTakeOnLine();
    // Look for the take from stash word
    if (take == null || !parser.lineHasWord(Keywords.TAKE_FROM_STASH_WORDS)) {
      return null;
    }
    int line = parser.peek().line();
    // Look for the dog
    String subject = findDogOnLine();
    if (subject == null) {
      throw new BarkError(line, "Who is taking from the jar? Name the dog or use she/he first");
    }
    // Edge case: "she whimpers as she wants the first biscuit from her cookie jar"
    List<AstNode> steps = new ArrayList<>();
    if (lineHasPrintVerb()) {
      PrintStyle style = findPrintStyleOnLine();
      steps.add(
          new AstNode.Print(
              style != null ? style : PrintStyle.BARK,
              List.of(new ParseExpression.Empty()),
              line,
              false,
              findDogSpeakerOnLine()));
    }
    steps.add(new AstNode.TakeFromStash(subject, take.stash(), take.which(), line));
    parser.consumeUntilLineEnd();
    if (steps.size() == 1) {
      return steps.get(0);
    }
    return new AstNode.LineGroup(steps, line);
  }

  // Find "from her cookie jar" on the line, then read backwards for first/last or the item
  private StashTake findStashTakeOnLine() {
    // Where this line starts and ends in the token list
    int lineStart = parser.cursorIndex();
    int lineEnd = parser.lineEndIndex();
    List<Token> tokens = parser.allTokens();
    // Walk each token on the line until we find a from word
    for (int fromAt = lineStart; fromAt < lineEnd; fromAt++) {
      Token token = tokens.get(fromAt);
      // Skip filler words
      if (!token.is(TokenType.IDENTIFIER)) {
        continue;
      }
      // Skip words that are not from
      if (!Keywords.FROM_WORDS.contains(parser.normalise(token.value()))) {
        continue;
      }
      // Read the registered stash name after from (example: her cookie jar becomes cookie_jar)
      String stash = ConfigLoader.resolveCollectionFromTokens(tokens, fromAt + 1, lineEnd);
      if (stash == null || !ConfigLoader.isStash(stash)) {
        continue;
      }
      // Read backwards from 'from' for words like first or "bone"
      String which = findWhichBeforeFrom(tokens, lineStart, fromAt);
      if (which != null) {
        return new StashTake(stash, which);
      }
      return null;
    }
    return null;
  }

  private String findWhichBeforeFrom(
      final List<Token> tokens, final int lineStart, final int fromAt) {
    // Go left from 'from', looking for the item the dog wants
    for (int i = fromAt - 1; i >= lineStart; i--) {
      Token token = tokens.get(i);
      // Name on the line (example: "bone")
      if (token.is(TokenType.STRING)) {
        return token.value();
      }
      // Number item on the line (example: '3')
      if (token.is(TokenType.NUMBER)) {
        return parser.normalise(token.value());
      }
      if (!token.is(TokenType.IDENTIFIER)) {
        continue;
      }
      String word = parser.normalise(token.value());
      // Spot word (example: first, second, last)
      if (StashSpots.isStashSpot(word)) {
        return word;
      }
    }
    return null;
  }

  private String findDogOnLine() {
    for (int offset = 0; !parser.isAtEndOrLineEndOffset(offset); offset++) {
      Token token = parser.peekAt(offset);
      if (!token.is(TokenType.IDENTIFIER)) {
        continue;
      }
      String word = parser.normalise(token.value());
      if (Keywords.isDogSubjectReference(word, token.value())) {
        return word;
      }
    }
    return null;
  }

  private String findDogSpeakerOnLine() {
    return findDogOnLine();
  }

  private boolean lineHasPrintVerb() {
    for (int offset = 0; !parser.isAtEndOrLineEndOffset(offset); offset++) {
      Token token = parser.peekAt(offset);
      if (token.is(TokenType.IDENTIFIER)
          && Keywords.PRINT_KEYWORDS.contains(parser.normalise(token.value()))) {
        return true;
      }
    }
    return false;
  }

  private PrintStyle findPrintStyleOnLine() {
    for (int offset = 0; !parser.isAtEndOrLineEndOffset(offset); offset++) {
      Token token = parser.peekAt(offset);
      if (token.is(TokenType.IDENTIFIER)) {
        String word = parser.normalise(token.value());
        if (Keywords.PRINT_KEYWORDS.contains(word)) {
          return Keywords.PRINT_STYLES.getOrDefault(word, PrintStyle.BARK);
        }
      }
    }
    return null;
  }
}

File: pbark/pbark/parser/collection/take_parser.py
from __future__ import annotations

from dataclasses import dataclass

from pbark.config.config_loader import ConfigLoader
from pbark.errors import BarkError
from pbark.lexer import TokenType
from pbark.parser import ast_node as AstNode
from pbark.parser import keywords as Keywords
from pbark.parser.collection import stash_spots as StashSpots
from pbark.parser.parse_expression import Empty
from pbark.print_style import PrintStyle


@dataclass
class _StashTake:
    stash: str
    which: str


class TakeParser:
    def __init__(self, parser) -> None:
        self._parser = parser

    def parse_take_line(self):
        take = self._find_stash_take_on_line()
        if take is None or not self._parser.line_has_word(Keywords.TAKE_FROM_STASH_WORDS):
            return None
        line = self._parser.peek().line
        subject = self._find_dog_on_line()
        if subject is None:
            raise BarkError(line, "Who is taking from the jar? Name the dog or use she/he first")
        steps = []
        if self._line_has_print_verb():
            style = self._find_print_style_on_line()
            steps.append(
                AstNode.Print(
                    style if style is not None else PrintStyle.BARK,
                    [Empty()],
                    line,
                    False,
                    self._find_dog_speaker_on_line(),
                )
            )
        steps.append(AstNode.TakeFromStash(subject, take.stash, take.which, line))
        self._parser.consume_until_line_end()
        if len(steps) == 1:
            return steps[0]
        return AstNode.LineGroup(steps, line)

    def _find_stash_take_on_line(self) -> _StashTake | None:
        line_start = self._parser.cursor_index()
        line_end = self._parser.line_end_index()
        tokens = self._parser.all_tokens()
        for from_at in range(line_start, line_end):
            token = tokens[from_at]
            if not token.is_(TokenType.IDENTIFIER):
                continue
            if self._parser.normalise(token.value) not in Keywords.FROM_WORDS:
                continue
            stash = ConfigLoader.resolve_collection_from_tokens(tokens, from_at + 1, line_end)
            if stash is None or not ConfigLoader.is_stash(stash):
                continue
            which = self._find_which_before_from(tokens, line_start, from_at)
            if which is not None:
                return _StashTake(stash, which)
            return None
        return None

    def _find_which_before_from(self, tokens, line_start: int, from_at: int) -> str | None:
        for i in range(from_at - 1, line_start - 1, -1):
            token = tokens[i]
            if token.is_(TokenType.STRING):
                return token.value
            if token.is_(TokenType.NUMBER):
                return self._parser.normalise(token.value)
            if not token.is_(TokenType.IDENTIFIER):
                continue
            word = self._parser.normalise(token.value)
            if StashSpots.is_stash_spot(word):
                return word
        return None

    def _find_dog_on_line(self) -> str | None:
        offset = 0
        while not self._parser.is_at_end_or_line_end_offset(offset):
            token = self._parser.peek_at(offset)
            if not token.is_(TokenType.IDENTIFIER):
                offset += 1
                continue
            word = self._parser.normalise(token.value)
            if Keywords.is_dog_subject_reference(word, token.value):
                return word
            offset += 1
        return None

    def _find_dog_speaker_on_line(self) -> str | None:
        return self._find_dog_on_line()

    def _line_has_print_verb(self) -> bool:
        offset = 0
        while not self._parser.is_at_end_or_line_end_offset(offset):
            token = self._parser.peek_at(offset)
            if token.is_(TokenType.IDENTIFIER) and self._parser.normalise(token.value) in Keywords.PRINT_KEYWORDS:
                return True
            offset += 1
        return False

    def _find_print_style_on_line(self) -> PrintStyle | None:
        offset = 0
        while not self._parser.is_at_end_or_line_end_offset(offset):
            token = self._parser.peek_at(offset)
            if token.is_(TokenType.IDENTIFIER):
                word = self._parser.normalise(token.value)
                if word in Keywords.PRINT_KEYWORDS:
                    return Keywords.PRINT_STYLES.get(word, PrintStyle.BARK)
            offset += 1
        return None
